external_samples/component.py

```diff
@@ -0,0 +1,68 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from collections.abc import Callable, Protocol
+
+class EmptyCallable(Protocol):
+    def __call__(self) -> None:
+        pass
+class PortType(Enum):
+    CAN_PORT = 1
+    SMART_IO_PORT = 2
+    SMART_MOTOR_PORT = 3
+    SERVO_PORT = 4
+    I2C_PORT = 5
+    USB_PORT = 6
+
+class InvalidPortException(Exception):
+    pass
+
+# This is an abstract class
+class Component:
+    @abstractmethod
+    def __init__(self, ports : list[tuple[PortType, int]]):
+        pass
+    # This is the manufacturer of the component
+    @abstractmethod
+    def get_manufacturer(self) -> str:   
+        pass 
+    # This is the name of the component
+    @abstractmethod
+    def get_name(self) -> str:  
+        pass
+    # This is the part number of the component
+    @abstractmethod
+    def get_part_number(self) -> str:  
+        pass
+    # This is the URL of the component
+    @abstractmethod
+    def get_url(self) -> str:  
+        pass
+    # This is the version of the software (returned as a (major, minor, revision) tuple where 
+    # major and minor are positive integers
+    # revision is an optional string
```

Comment:
Does optional mean that the returned tuple can have just 2 items? Or that the last item might be an empty string or None?

Reply:
That the last item might be an empty string.   Let me change the comment.